src/tbb/arena.h

```diff
@@ -179,6 +179,74 @@ class atomic_flag {
     }
 };
 
+#if __TBB_PREVIEW_PARALLEL_PHASE
+class thread_leave_manager {
+    static const std::uintptr_t DELAYED_LEAVE       = 0;
+    static const std::uintptr_t FAST_LEAVE          = 1;
+    static const std::uintptr_t ONE_TIME_FAST_LEAVE = 1 << 1;
+    static const std::uintptr_t PARALLEL_PHASE      = 1 << 2;
+
+    std::atomic<std::uintptr_t> my_state{UINTPTR_MAX};
+public:
+    // This method is not thread-safe!
+    // Required to be called after construction to set initial state of the state machine.
+    void set_initial_state(tbb::task_arena::leave_policy lp) {
+        if (lp == tbb::task_arena::leave_policy::automatic) {
+            std::uintptr_t platform_policy = governor::hybrid_cpu() ? FAST_LEAVE : DELAYED_LEAVE;
+            my_state.store(platform_policy, std::memory_order_relaxed);
+        } else {
+            __TBB_ASSERT(lp == tbb::task_arena::leave_policy::fast,
+                         "Was the new value introduced for leave policy?");
+            my_state.store(FAST_LEAVE, std::memory_order_relaxed);
+        }
+    }
+
+    void reset_if_needed() {
+        std::uintptr_t curr = my_state.load(std::memory_order_relaxed);
+        if (curr == ONE_TIME_FAST_LEAVE) {
+            // Potentially can override decision of the parallel phase from future epoch
+            // but it is not a problem because it does not violate the correctness
+            my_state.fetch_and(~ONE_TIME_FAST_LEAVE);
+        }
+    }
+
+    // Indicate start of parallel phase in the state machine
+    void register_parallel_phase() {
+        __TBB_ASSERT(my_state.load(std::memory_order_relaxed) != UINTPTR_MAX, "The initial state was not set");
+
+        std::uintptr_t prev = my_state.fetch_add(PARALLEL_PHASE);
+        __TBB_ASSERT(prev + PARALLEL_PHASE > prev, "Overflow detected");
+        if (prev & ONE_TIME_FAST_LEAVE) {
+            // State was previously transitioned to "One-time Fast leave", thus with the start
+            // of new parallel phase, it should be reset
+            my_state.fetch_and(~ONE_TIME_FAST_LEAVE);
+        }
+    }
+
+    // Indicate the end of parallel phase in the state machine
+    void unregister_parallel_phase(bool enable_fast_leave) {
+        std::uintptr_t prev = my_state.load(std::memory_order_relaxed);
+        __TBB_ASSERT(prev != UINTPTR_MAX, "The initial state was not set");
+
+        std::uintptr_t desired{};
+        do {
+            __TBB_ASSERT(prev - PARALLEL_PHASE < prev,
+                         "A call to unregister without its register complement");
+            desired = prev - PARALLEL_PHASE; // Mark the end of this phase in reference counter
+            if (enable_fast_leave && /*it was the last parallel phase*/desired == DELAYED_LEAVE) {
```

Comment:
A bit of nitpicking: the comment does not really match the condition tested. To check if that's the last parallel phase, you'd use `desired < PARALLEL_PHASE`. Here you also check that neither of the fast-leave flags is set. Perhaps can be improved like this:
```suggestion
            if (enable_fast_leave && /*it was the last parallel phase, and delayed leave is the default*/ desired == DELAYED_LEAVE) {
```